src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -0,0 +1,133 @@
+    between runs, as opposed to the 'live' tables that exist during an update.
+
+    Modified entries are similar to added inputs, but with a subtle difference. When an input is Added
+    all outputs are unconditionally added too. However when an input is modified, the outputs may still
+    be the same (for instance something changed elsewhere in a file that had no bearing on the produced
+    output). In this case, the state table checks the results against the previously produced values,
+    and any that are found to be the same instead get a cached state, meaning no new downstream work 
+    will be produced for them. Thus a modified input is the only slot that can have differing output 
+    states.
+    */
+
+    internal enum EntryState { Added, Removed, Modified, Cached };
+
+    internal interface IStateTable
+    {
+        IStateTable Compact();
+    }
+
+    /// <summary>
+    /// A data structure that tracks the inputs and output of an execution node
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal sealed class NodeStateTable<T> : IStateTable
+    {
+        internal static NodeStateTable<T> Empty { get; } = new NodeStateTable<T>(ImmutableArray<ImmutableArray<(T, EntryState)>>.Empty, isCompacted: true, exception: null);
+
+        //PROTOTYPE: there is no need to store the state per item
+        //           we can instead store one state per input, with
+        //           an optional set for modified states.
+        readonly ImmutableArray<ImmutableArray<(T item, EntryState state)>> _states;
+
+        private readonly bool _isCompacted;
+
+        readonly Exception? _exception;
+
+        private NodeStateTable(ImmutableArray<ImmutableArray<(T, EntryState)>> states, bool isCompacted, Exception? exception)
+        {
+            _states = states;
+            _isCompacted = isCompacted;
+            _exception = exception;
+        }
+
+        public bool IsFaulted { get => _exception is not null; }
+
+        public IEnumerator<(T item, EntryState state)> GetEnumerator()
+        {
+            return _states.SelectMany(s => s).GetEnumerator();
+        }
+
+        public IStateTable Compact()
+        {
+            if (_isCompacted)
+                return this;
+
+            var compacted = ArrayBuilder<ImmutableArray<(T, EntryState)>>.GetInstance();
+            foreach (var entry in _states)
+            {
+                // we have to keep empty entries
+                // we only remove all entries at once, so only need to check the first item
+                if (entry.Length == 0 || entry[0].state != EntryState.Removed)
+                {
+                    compacted.Add(entry.SelectAsArray(e => (e.item, EntryState.Cached)));
+                }
+            }
+            return new NodeStateTable<T>(compacted.ToImmutableAndFree(), isCompacted: true, _exception);
+        }
+
+        public static NodeStateTable<T> FromFaultedTable<U>(NodeStateTable<U> table)
```

Comment:
>FromFaultedTable<U> [](http://example.com/codeflow?start=40&length=19)

This method doesn't appear to be used. Consider removing or adding a PROTOTYPE comment.